src/main/java/com/google/sps/servlets/queue/EntityListServlet.java

```
@@ -0,0 +1,175 @@
+     * @return: Stringified JSON request
+     * @throws IOException: if the body can't be parsed
+     */
+    private String parseBody(HttpServletRequest request) throws IOException {
+        return request.getReader().lines()
+                .reduce("", (accumulator, actual) -> accumulator + actual);
+    }
+
+    /**
+     * A helper function to get entity data from database
+     * @param type: what list type should it look for (queue / watched)
+     * @param userID: userID to look for
+     * @return: a list of abstracted type EntityDbQueue
+     */
+    private List<? extends EntityDbQueue> getQueueWithType(String type, String userID) {
+        if(type.equals(typeQueue)) {
+            return ofy().load().type(WantToWatchQueueObject.class).filter("userID", userID).list();
+        } else {
+            return ofy().load().type(WatchedQueueObject.class).filter("userID", userID).list();
+        }
+    }
+
+    /**
+     * helper function to check if the type is valid
+     * @param queueType: type to check
+     * @return: true / false if the type is valid
+     */
+    private boolean isCorrectListType(String queueType) {
+        return queueType.equals(typeQueue) || queueType.equals(typeViewed);
+    }
+
+    /**
+     * Used to decide which data type we should map POST request to
+     * @param body: body to parse into object
+     * @return: null if invalid entity type, an EntityDbQueue type object if it can
+     * @throws JsonProcessingException: if it can't parse out data.
+     */
+    private EntityDbQueue decideDbType(String body) throws JsonProcessingException {
+        EntityDbQueue entity = mapper.readValue(body, EntityDbQueue.class);
+        if(entity.getEntityType().equals(typeViewed)) {
+            return mapper.readValue(body, WatchedQueueObject.class);
+        } else if(entity.getEntityType().equals(typeQueue)) {
+            return mapper.readValue(body, WantToWatchQueueObject.class);
+        } else {
+            return null;
```

Instead of a null, is it not better to throw an exception and catch it?

---

sounds like a good idea!

---

Just fixed with 21d6cdb